### ISSUE TITLE
Parallel processing of paired-end fastq files

### ISSUE THREAD

**Issue by PramodRaoB** [FelixKrueger/TrimGalore#195]
From my understanding of the code, in paired-end processing, both fastq input files could be processed parallelly, but they are currently sequential. Is this correct? If so, I could take this feature up.

**Comment by FelixKrueger** [comment]
Yes, it this is correct. If the trimming could occur in parallel it might indeed speed up the run-time, which sill allowing the the same flexibility further downstream. It might have to be taken care of for resource allocation later on, e.g. on nf-core, but this would be a subsequent step. If you want to have a go at it - that could be nice!

**Comment by PramodRaoB** [comment]
I was thinking of a different workflow that would reduce the execution time by removing some I/O overhead.

Process both of the paired-end files parallelly by cutadapt and have both the outputs streamed. Then, combine the trimming and validation phase and write the final output to disk directly. Since Cutadapt scales well for thread count less than 8, processing them in parallel (with half the original thread-count) would still give the same effective runtime. But, this would avoid writing the intermediary trimmed outputs (which are potentially compressed leading to even higher runtimes) thereby giving a lower overall runtime.

Let me know if this sounds good @FelixKrueger ! Thanks.

**Comment by c-mertes** [comment]
Dear @PramodRaoB  any progress on this? As I think one could easily reduce the processing time and also the resource allocation by just running it in parallel.

The easy step, I guess would be just forking the processes and wait for them to finish. A more hands on solution, could be the solution mentioned by @PramodRaoB.

**Comment by FelixKrueger** [comment]
As a quick update here: I am in the process of releasing a rewrite of Trim Galore very soon, which should also address and fix this issue. See the beta release announcement here: https://github.com/FelixKrueger/TrimGalore/discussions/215

**Comment by FelixKrueger** [comment]
Closing this with an update for context, but please reopen if v2 doesn't actually address your case.

The Rust rewrite (v2.x — currently in beta, GA targeting 2026-05-04) restructures paired-end processing into a single-pass worker pool: both R1 and R2 are trimmed and gzip-compressed concurrently in worker threads, with per-pair adapter auto-detection. There is no longer a separate "trim then validate" stage with intermediate writes — it's one streaming pass per pair. `--cores N` controls the worker count and gives near-linear speedup up to ~16 cores (typically I/O-bound beyond that).

So the original ask here — "process both files in parallel without writing intermediates" — is the v2 default; the architectural overhead this issue points at is gone in v2.

Try it via:

```bash
cargo install trim-galore --version 2.1.0-beta.7
# or
docker run --rm -v "$PWD":/data -w /data ghcr.io/felixkrueger/trimgalore:beta \
    trim_galore --cores 8 --paired R1.fastq.gz R2.fastq.gz
```

If your particular workflow still has a parallelism gap that v2 doesn't cover, please reopen — happy to look at it concretely.

Thanks @PramodRaoB for the original report and @c-mertes for keeping it on the radar.

— Felix
